[Step3]인증을 통한 기능 구현 PR 요청 드립니다.

re-mora opened the pull request:
안녕하세요 :) 교육생 최용석입니다 😄

3단계 `인증을 통한 기능 구현` 미션 구현이 완료되어, 리뷰 요청드립니다.

미션에 주어진 요구사항을 step3-README.md 문서에 정리 후 아래와 같이 3가지 파트로 나누어서 기능을 구현해 보았습니다.
- 토큰 발급 인수 테스트
- 내 정보 조회 기능
- 즐겨 찾기 기능 구현

이번 미션 역시 미흡한 부분이 있다면 피드백 부탁드립니다 :)

re-mora committed:
docs: 요구사항 및 TODO List 작성

re-mora committed:
test: 토큰 발급 인수 테스트 구현

- 토큰 발급(로그인) 기능 구현 (로그인 요청 : email, password)
- 주어진 `AuthAcceptanceTest`통과 하도록 구현
- 주어진 `AuthAcceptanceTest`의 예외 케이스 구현
- 인수테스트 작성 편의를 위한 RestAssuredUtils 작성

구현된 기능의 TODO List 갱신

re-mora committed:
test: 내 정보 관련 인수테스트 구현

- 내 정보 조회 : GET `/members/me`
- 내 정보 수정 : PUT `/members/me`
- 내 정보 삭제 : DELETE `/members/me`
- 잘못된 토큰 정보로 내 정보 관련 API 요청 시 예외 케이스 구현

구현된 기능의 TODO List 갱신

re-mora committed:
test: 즐겨 찾기 관리 기능 관련 인수테스트 구현

- 즐겨 찾기 생성/목록 조회/삭제 관련 요청/검증 부 구현
- 즐겨 찾기 API 요청/응답 객체 정의
- 즐겨 찾기 Entity 정의

re-mora committed:
feat: 즐겨 찾기 관리 Service 및 TC 작성

re-mora committed:
refactor: 즐겨찾기 엔티티에 누락된 회원과의 연관관계 매핑 추가

re-mora committed:
refactor: 즐겨찾기 엔티티에 누락된 회원과의 연관관계 매핑 추가

즐겨찾기 목록 조회, 즐겨찾기 삭제 쿼리 조건에 누락된 회원 파라미터 추가

즐겨찾기 목록 조회 시, FetchType.Lazy로 설정된 연관관계의 프록시 객체 초기화 과정으로 인해 N+1문제가 발생하는 기존 쿼리에 EntityGraph를 적용하여 즉시 조회

re-mora committed:
feat: 즐겨찾기 관련 기능 요청 수신 부인 Controller 구현

re-mora committed:
test: 잘못된 토큰 정보로 즐겨찾기 관련 APU 요청시 예외 발생 검증 인수 테스트 작성

re-mora committed:
chore(comment-polish): 인수 조건 주석 정리

re-mora committed:
Merge branch 'mission/step3-preview' into mission/step3

re-mora committed:
chore(comment-polish): 즐겨 찾기 관련 테스트 package 경로 수정

dhmin5693 commented on src/main/java/nextstep/subway/favorites/application/FavoritesService.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    @Transactional
+    public void deleteFavoritesById(LoginMember loginMember, Long id) {
```
이 메소드의 단위 테스트가 없습니다.

re-mora replied:
누락된 단위 테스트와, 즐겨찾기 삭제에 대한 인수테스트도 같이 작성해 보겠습니다!

dhmin5693 commented on src/main/java/nextstep/subway/favorites/dto/FavoritesRequest.java:
```diff
@@ -0,0 +1,31 @@
+public class FavoritesRequest {
+
+    private Long sourceId;
+    private Long targetId;
```
미션 요구사항과 스펙이 다르네요. 확인해주세요.

```json
{
    "source": "1",
    "target": "3"
}
```

re-mora replied:
ㅠㅠ.. 요구 사항을 자꾸 누락하는 모습을 보여드리네요.. 😭 
반성하면서 요구 사항에 맞도록 파라미터명을 수정하겠습니다.

꼼꼼하게 봐주셔서 다시 한번 감사드립니다!

dhmin5693 commented on src/main/java/nextstep/subway/favorites/application/FavoritesService.java:
```diff
@@ -0,0 +1,67 @@
+
+        Favorites persistFavorites = favoritesRepository.save(favorites);
+        return FavoritesResponse.from(persistFavorites);
+    }
```
특정 사용자가 이미 등록한 노선에 대해서는 중복 등록을 방지해야 하지 않을까요?

re-mora replied:
앗! 말씀하신 경우를 생각하지 못했네요.. 😭 

이미 존재하는 구간인지를 확인하는 로직을 추가하도록 하겠습니다!

dhmin5693 commented on src/main/java/nextstep/subway/favorites/domain/FavoritesRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+public interface FavoritesRepository extends JpaRepository<Favorites, Long> {
+
+    @EntityGraph(attributePaths = {"source", "target"})
+    List<Favorites> findAllByMemberId(Long memberId);
```
👍

dhmin5693 commented on src/main/java/nextstep/subway/auth/application/AuthService.java:
```diff
@@ -29,7 +29,7 @@ public TokenResponse login(TokenRequest request) {
     public LoginMember findMemberByToken(String credentials) {
         if (!jwtTokenProvider.validateToken(credentials)) {
-            return new LoginMember();
+            throw new AuthorizationException(AUTHORIZATION_ERROR);
```
👍

dhmin5693 reviewed (changes requested):
안녕하세요.
3단계 미션 매우 잘해주셨군요.

요청 스펙이 안맞는 부분이 있는데 여기만 빠르게 확인해주세요. 🙏

re-mora committed:
fix: 잘못 명시된 요구사항 API 스펙의 요청 파라미터명 수정

요구사항
```
POST /favorites HTTP/1.1
{
    "source": "1",
    "target": "3"
}
```

오작성
POST /favorites HTTP/1.1
```
POST /favorites HTTP/1.1
{
    "sourceId": "1",
    "targetId": "3"
}
```

re-mora committed:
fix: 즐겨찾기 추가 시, 이미 등록된 구간인 경우 예외 처리 추가 및 테스트 작성

등록 대상 즐겨찾기 존재 여부 확인 JPA 쿼리 메소드 추가

대상 구간이 이미 등록된 즐겨찾기 인 경우 예외 발생 테스트 코드 작성
- 서비스 계층에 대한 TC 작성
- API에 대한 인수테스트 작성

즐겨찾기 서비스 계층 테스트 코드를 위한 객체 생성 부 리팩토링

반복적인 ExceptionHandler 추출

re-mora committed:
test: 즐겨찾기 삭제 기능에 대한 누락된 TC 작성

서비스 계층 및 인수테스트 추가

re-mora commented:
안녕하세요. 동휘님 :)

말씀하셨던 부분을 아래와 같이 수정하였습니다.

- 잘못 명시된 요구사항 API 스펙의 요청 파라미터명 수정
- 즐겨찾기 대상 구간이 이미 존재하는 경우에 대한 예외 처리
- 누락된 테스트 케이스 작성

작성한 내용에 미흡한 부분이 있다면 한번 더 피드백 부탁드립니다!

---
미션 막바지에 다가가고 있는데 항상 꼼꼼하게 리뷰해 주셔서 감사합니다!
날씨가 점점 더워지는데, 항상 건강 조심하시고
좋은 주말 보내세요 😄

dhmin5693 commented on src/main/java/nextstep/subway/favorites/application/FavoritesService.java:
```diff
@@ -0,0 +1,75 @@
+        return FavoritesResponse.from(persistFavorites);
+    }
+
+    private boolean isExistFavorites(Long sourceId, Long targetId) {
+        return favoritesRepository.existsBySourceIdAndTargetId(sourceId, targetId);
+    }
```
이러면 다른 사용자가 같은 노선을 등록해도 오류가 날거에요.
차라리 member_id, source_id, target_id 를 유니크 인덱스로 묶어버리는건 어떻게 생각하세요?
그리고 `favoritesRepository.save`를 `@DataJpaTest` 로 검증해주면 될거같아요.

dhmin5693 reviewed (approved):
요구사항을 항상 꼼꼼히 확인해주시기 바라요 😄
이번 미션에서 요구하는 학습을 충분히 마치셔서 merge하도록 하겠습니다.

하나 더 코멘트를 남겨두었는데요, 충분히 생각해보시고 다음 미션에 반영 여부를 결정해주세요!
